### PR TITLE
chain: add mempool lookup for outpoints

### PR DESCRIPTION
**Pull request by yyforyongyu**
This PR adds a new method `LookupInputMempoolSpend` to directly check whether a given input has been spent or not in the mempool.

### CODE REVIEW

File: chain/bitcoind_events_test.go [hunk 1]
```diff
@@ -214,6 +217,46 @@ func testNotifySpentMempool(t *testing.T, miner *rpctest.Harness,
 	}
 }
 
+// testLookupInputMempoolSpend tests that LookupInputMempoolSpend returns the
+// correct tx hash and whether the input has been spent in the mempool.
+func testLookupInputMempoolSpend(t *testing.T, miner *rpctest.Harness,
+	client *BitcoindClient) {
+
+	require := require.New(t)
```
Comment (Roasbeef) [inline chain/bitcoind_events_test.go:225]: Maybe call it like `rt` to avoid shadowing the import path?

File: chain/bitcoind_events_test.go [hunk 2]
```diff
@@ -214,6 +217,46 @@ func testNotifySpentMempool(t *testing.T, miner *rpctest.Harness,
 	}
 }
 
+// testLookupInputMempoolSpend tests that LookupInputMempoolSpend returns the
+// correct tx hash and whether the input has been spent in the mempool.
+func testLookupInputMempoolSpend(t *testing.T, miner *rpctest.Harness,
```
Comment (Roasbeef) [inline chain/bitcoind_events_test.go:222]: Can we refactor these tests to also exercise the `btcd` version? Or is this already covered by the `btcd` integration tests themselves?

File: chain/bitcoind_events_test.go (hunk 1 above)
Comment (yyforyongyu) [reply chain/bitcoind_events_test.go:225]: Done.

File: chain/bitcoind_events_test.go (hunk 2 above)
Comment (yyforyongyu) [reply chain/bitcoind_events_test.go:222]: Yeah that's covered there, but I definitely agree we should add a new test `btcd_test.go` here to check the RPC client behavior, maybe something to do when we decide to refactor the `chain` package.

**Comment by saubyk** [comment]
cc: @guggero @Crypt-iQ for review please

File: chain/bitcoind_zmq_events.go [hunk 3]
```diff
@@ -501,28 +484,7 @@ func (b *bitcoindZMQEvents) mempoolPoller() {
...
-	resp, err := client.RawRequest(
-		"gettxspendingprevout", []json.RawMessage{req},
-	)
-	if err != nil {
-		return chainhash.Hash{}, false
-	}
-
-	var prevoutResps []getTxSpendingPrevOutResp
-	err = json.Unmarshal(resp, &prevoutResps)
+	prevoutResps, err := client.GetTxSpendingPrevOut([]wire.OutPoint{op})
```
Comment (Chinwendu20) [inline chain/bitcoind_zmq_events.go:487]: Nice.

File: chain/btcd.go [hunk 4]
```diff
@@ -91,6 +93,109 @@ func NewRPCClient(chainParams *chaincfg.Params, connect, user, pass string, cert
...
+// NewRPCClientWithConfig creates a client connection to the server based on
+// the config options supplised.
+//
+// The connection is not established immediately, but must be done using the
+// Start method.  If the remote server does not operate on the same bitcoin
+// network as described by the passed chain parameters, the connection will be
+// disconnected.
+func NewRPCClientWithConfig(cfg *RPCClientConfig) (*RPCClient, error) {
+	// Make sure the config is supplied.
+	if cfg == nil {
```
Comment (guggero) [inline chain/btcd.go:151]: nit: the way `Go` works, this check could be in the `validate()` method.

**Review by guggero — APPROVED** [review]
utACK, LGTM :tada:

**Comment by Roasbeef** [comment]
Can be rebased with the `btcd` version merged!

File: chain/btcd.go (hunk 4 above)
Comment (yyforyongyu) [reply chain/btcd.go:151]: Won't that cause a nil deference? Or you mean before initializing the client we do a validation on config in `lnd`?

Comment (guggero) [reply chain/btcd.go:151]: No, the thing is, in Go you can call a method on a struct, even if the struct itself is nil.
So you can do:
```
// validate checks the required config options are set.
func (r *RPCClientConfig) validate() error {
	// Make sure the config is supplied.
	if r == nil {
		return errors.New("missing rpc config")
	}

	// Make sure retry attempts is positive.
	if r.ReconnectAttempts < 0 {
		return errors.New("reconnectAttempts must be positive")
	}
	...
```

Comment (yyforyongyu) [reply chain/btcd.go]: Ohhh cool, TILed, thanks!

Comment (Roasbeef) [reply chain/btcd.go]: Yeh at time a check on the nil pointer can actually be useful.

**Review by Roasbeef — APPROVED** [review]
LGTM 🎢

**Pull request merged**